### PR TITLE
Feat/298 implement anonymous session for intent preservation for questionnaire response resource

### LINKED ISSUE #298: Implement Guest Session and Regulated Ownership Transfer for QuestionnaireResponse
# User Story

As an anonymous user, I want to fill in questionnaires without logging in and later reclaim my submitted data after registering, so that my responses are preserved and securely associated with my account.

# Description

Implement a complete guest session workflow using SuperTokens that allows anonymous users to create QuestionnaireResponses and later transfer ownership to an authenticated user upon registration.

The backend already supports anonymous sessions via `POST /auth/anonymous-session`, guest role assignment, and unique anonymous session identifiers. This task extends the existing infrastructure to ensure proper session-based data isolation, secure ownership transfer, and cleanup after registration.

QuestionnaireResponses created during a guest session must be associated with the anonymous session identifier. During user registration, the system must validate session ownership and migrate only the relevant QuestionnaireResponses to the newly authenticated user, preventing cross-user data access.

# Reference

- Existing endpoint: `POST /auth/anonymous-session`
- SuperTokens Anonymous Session & Ownership Transfer  
  https://supertokens.com/docs/post-authentication/session-management/advanced-workflows/anonymous-session#3-transfer-the-data-to-a-logged-in-session
- Current QuestionnaireResponse creation and ownership logic
- SuperTokens `SessionOptional` middleware usage

**Notes:**

- Store the anonymous session identifier in the QuestionnaireResponse `identifier` field at creation time.
- Ensure strict isolation so guest sessions cannot access each other’s data.
- Extend the registration flow to accept and validate a pre-auth anonymous session identifier.
- Implement a service to transfer QuestionnaireResponse ownership from anonymous session to authenticated user.
- Validate session ownership before transfer to prevent hijacking or cross-user access.
- Update FHIR resource references during ownership transfer as needed.
- Revoke and clean up anonymous sessions after successful registration.
- Handle edge cases such as tab close, expired sessions, or abandoned guest sessions.
- Add logging and tests to cover:
  - Multiple concurrent guest sessions
  - Successful ownership transfer
  - Failed or invalid transfer attempts

### PR DESCRIPTION
**Pull request by luckyAkbar**
## Summary

Rework the existing anonymous session generation endpoint to be able to create JWT token for each guest embedded with unique guest-id to be used by frontend to uniquely identify any resources created by a guest. Also opens up a new endpoint to serve as a resource claim to help preserve user intention by claiming a set of known resources created by a particular guest.

## Purpose

This PR attempts to solve [this issue](https://github.com/konsulin-care/konsulin-api/issues/298).

## Design Decision

1. Anonymous Session Generation Endpoint

This endpoint will only mint a new token if the requester didn't already have a valid guest token. Thus, there is a possibility when frontend call this function but the backend didn't generate any new token, instead the token will only parse the embedded `guest-id` and return it back to the frontend. This was made because guest token in frontend is saved in a cookie with `httpOnly` enabled. That way, the token will be protected by Xss, but the frontend wouldn't be able to parse and read the embedded guest-id. Thus, calling this endpoint again will help frontend to know the value again

2. Resources Claim Endpoint

This endpoint will only allow a user to claim any resources he made while still a guest if and only if the user can present both valid guest token and supertoken session (after logged in), that way we can ensure only a registered user can claim the resources generated during guest session.

## Testing

This feature has been tested using these scenarios detailed below:

### Case 1: generating new anonymous session without any existing anonymous session

(the token is new, indicated by `"is_new": true,`)
image

### Case 2: calling the anonymous session generation endpoint again with existing anonymous session token

(the token is not new, indicated by `"is_new": false`)
image

### Case 3: Claiming resources created during guest session

image
image

---

<a href="https://app.devin.ai/review/konsulin-care/konsulin-api/pull/300">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

## Summary by CodeRabbit

* **New Features**
  * Anonymous session support with cookie-based guest tokens, guest_id, and is_new flag; supports a force_new option.
  * Endpoint to claim anonymous resources into an authenticated account; returns claimed status, count, and reference list; clears anon cookie on success.

* **Infrastructure**
  * CORS updated to allow PATCH requests; new PATCH route for anonymous-claim added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by luckyAkbar** [comment]
Currently this PR [got flagged by codacy ](https://github.com/konsulin-care/konsulin-api/pull/300/checks?check_run_id=61874508164) because of `A session cookie was detected without setting the 'Secure' flag.`

The problem still persist even after I explicitly set the `Secure` flag to `True` by default. I'll look at it again to find the solution. Other than that, this PR should be ready for review mas @lamurian

**Comment by lamurian** [comment]
I noticed that the `Secure` field is set to a variable, which resolve as `true` unless application environment is set to `local`. Codacy failed to notice this pattern because it only check the code statically. For now, I've ignored and mark this flag as false positive @luckyAkbar.

File: internal/app/delivery/http/controllers/auth_controller.go [hunk 1]
```diff
@@ -260,18 +268,114 @@ func (ctrl *AuthController) CreateAnonymousSession(w http.ResponseWriter, r *htt
...
+		secure := true
+		if strings.EqualFold(ctrl.InternalConfig.App.Env, "local") {
+			secure = false
+		}
+		http.SetCookie(w, &http.Cookie{
+			Name:     constvars.AnonymousSessionCookieName,
+			Value:    result.Token,
+			Path:     "/",
+			Expires:  time.Now().Add(ttl),
+			MaxAge:   int(ttl.Seconds()),
```
Comment (lamurian) [inline internal/app/delivery/http/controllers/auth_controller.go]: I imagine configuring this as a persistent cookie will make guest management a bit challenging in a shared environment. Let's say that a clinic has a public computer for self-assessment purpose. A patient came in, then fill in the screening assessment. If this patient did not claim the resource and another patient claim the resource, won't this patient also get the previously unclaimed resource?

File: internal/app/delivery/http/controllers/auth_controller.go [hunk 2]
```diff
@@ -260,18 +268,114 @@ func (ctrl *AuthController) CreateAnonymousSession(w http.ResponseWriter, r *htt
...
+	secure := true
+	if strings.EqualFold(ctrl.InternalConfig.App.Env, "local") {
+		secure = false
+	}
+	http.SetCookie(w, &http.Cookie{
+		Name:     constvars.AnonymousSessionCookieName,
+		Value:    "",
+		Path:     "/",
+		Expires:  time.Unix(0, 0),
+		MaxAge:   -1,
```
Comment (lamurian) [inline internal/app/delivery/http/controllers/auth_controller.go:366]: I understand that this is a mechanism to delete the cookie after successful resource claim. However, won't it be better if we provide a "clear session" endpoint instead? That way, the frontend can flexibly determine when to clear the anonymous session and delete the associated cookie. In this scenario, cookie deletion may happen not only after a successful resource claim, but may also be on page refresh, tab closure, or other events detectable by the frontend.

Comment (lamurian) [reply internal/app/delivery/http/controllers/auth_controller.go:366]: It could be a `DELETE` operation to `/auth/anonymous/session`, so that frontend can control the session reset.

File: internal/app/services/core/auth/auth_usecase_impl.go [hunk 3]
```diff
@@ -355,21 +368,232 @@ func (uc *authUsecase) CreateMagicLink(ctx context.Context, request *requests.Su
...
+		)
+		return nil, err
+	}
+
+	entries := make([]map[string]any, 0, len(responses))
+	refs := make([]string, 0, len(responses))
+	for _, response := range responses {
+		if !canClaimQuestionnaireResponse(response, ownerRef) {
+			continue
+		}
```
Comment (lamurian) [inline internal/app/services/core/auth/auth_usecase_impl.go:461]: Shall we also implement a quick check to allow claiming within specific time window? I'm thinking about giving a 24-hour time window. Resource claiming will fail if it happens after the designated time window.

File: internal/pkg/constvars/anonymous_session.go [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package constvars
+
+const (
+	AnonymousSessionCookieName       = "anon_session"
+	AnonymousSessionGuestIDClaimKey  = "guest_id"
+	AnonymousSessionIdentifierSystem = "https://konsulin/anonymous-guest"
```
Comment (lamurian) [inline internal/pkg/constvars/anonymous_session.go]: Kindly change it to `https://login.konsulin.care/guestid` to ensure consistency with our SuperTokens identifier.

**Review by lamurian — CHANGES_REQUESTED** [review]

File: internal/app/services/core/auth/auth_usecase_impl.go (hunk 3 above)
Comment (luckyAkbar) [reply internal/app/services/core/auth/auth_usecase_impl.go:461]: We can indirectly control this by making the token itself only valid for 1 day, mas. That way, the backend will reject incoming requests with token that is more than 1 day old.

However, that effectively negates the supertokens anonymous session intent since [it recommends the token itself to be long lived](https://supertokens.com/docs/post-authentication/session-management/advanced-workflows/anonymous-session). 

But, we doesn't have to blindly follow that. Making the token only valid for 1 days has several benefits according to your need. The first is we don't have to implement custom check to filter any more than 1 days old resource to be claimed by the guest. second, the token will get refreshed and prevent old token being used more than once to claim resources.

Thus, I recommend making the token lived for only 1 day to cover that case. What do you think mas?

File: internal/app/delivery/http/controllers/auth_controller.go (hunk 1 above)
Comment (luckyAkbar) [reply internal/app/delivery/http/controllers/auth_controller.go]: Yes, that's true mas @lamurian. Currently, the token is saved in persistent cookie in the frontend and the only way the token can get removed is through signing in, thus making it possible for the second guest accidentally claim the first guest's resources. The solution to fix this will also relates to your inputs [here](https://github.com/konsulin-care/konsulin-api/pull/300#discussion_r2743887976).

However, I need your help to determine how the guest session get deleted other than the guest signing in after submitting a questionnaire. Should there be a button somewhere that will hit [this endpoint](https://github.com/konsulin-care/konsulin-api/pull/300#discussion_r2743896649) and remove the token, or something to trigger similar process elsewhere. After that, I might be able to offers several options to implement that, mas @lamurian

Comment (lamurian) [reply internal/app/delivery/http/controllers/auth_controller.go]: For now we don't have to provide a button to hit [that delete endpoint](https://github.com/konsulin-care/konsulin-api/pull/300#discussion_r2743896649). However, I'd like the frontend to call this endpoint upon page refresh.

File: internal/app/services/core/auth/auth_usecase_impl.go (hunk 3 above)
Comment (lamurian) [reply internal/app/services/core/auth/auth_usecase_impl.go:461]: I agree with you that we don't have strictly follow their recommendation of long-lived JWT for an anonymous session. The way I understood their intention is that, they provide a simple mechanism of anonymous session management so that it can be widely applicable to various scenarios. Personally, I think that their approach is a perfect fit for a native mobile application, where one user occupies one gadget. However, for our use case, modifying token validity to one day should suffice.

But again, won't that become a [persistent cookie](https://github.com/konsulin-care/konsulin-api/pull/300#discussion_r2743877629) with relatively shorter expiry period?Am I missing something or we can also configure expiry for a session cookie?

File: internal/app/delivery/http/controllers/auth_controller.go (hunk 1 above)
Comment (luckyAkbar) [reply internal/app/delivery/http/controllers/auth_controller.go]: To create the guest session deletion API and making the frontend calls the endpoint during manual page refresh is possible and the changes isn't complicated.

However, on which page should that be implemented? If we just make it happen on all page refresh, there is a possibility where a guest doing refresh the page showing the questionnaire result (before saving the result) and making the user unable to claim the QR resource.

I suggest, maybe we should only apply that to our home page, thus making every manual page refresh on homepage will trigger the delete endpoint and make such as it is a new guest that is accessing the app.

If that is your intention, please confirm that is the scenario you want, mas. That way, other than the guest signin to app, refreshing the page will also delete the guest token and issue a new one

**Comment by lamurian** [comment]
Agree Mas @luckyAkbar, implementing that only on the homepage sounds reasonable to me

File: internal/app/services/core/auth/auth_usecase_impl.go (hunk 3 above)
Comment (luckyAkbar) [reply internal/app/services/core/auth/auth_usecase_impl.go:461]: >But again, won't that become a https://github.com/konsulin-care/konsulin-api/pull/300#discussion_r2743877629 with relatively shorter expiry period?Am I missing something or we can also configure expiry for a session cookie?

Yes, that is right, mas. I choose to set `maxAge` and `expires` in the cookie to make it saved as persistent cookie instead of session cookie. However, if you want it to be treated as session cookie, I can make the changes and retest the whole feature. The downside using session cookie is when the browser closes, the guest will lost its guest token and unable to reclaim the resources

Comment (lamurian) [reply internal/app/services/core/auth/auth_usecase_impl.go:461]: > The downside using session cookie is when the browser closes, the guest will lost its guest token and unable to reclaim the resources

I think that's a pretty safe trade-off Mas. If a guest uses the app then closes the browser after exploring, most likely they're not compelled to register at that moment. Thus, deleting their session should be reasonable.

File: internal/app/delivery/http/controllers/auth_controller.go (hunk 1 above)
Comment (lamurian) [reply internal/app/delivery/http/controllers/auth_controller.go]: > I suggest, maybe we should only apply that to our home page, thus making every manual page refresh on homepage will trigger the delete endpoint and make such as it is a new guest that is accessing the app.
> 
> If that is your intention, please confirm that is the scenario you want, mas. That way, other than the guest signin to app, refreshing the page will also delete the guest token and issue a new one

I like this approach Mas, and I think deleting session upon refresh on the homepage is quite sensible.

File: internal/app/services/core/auth/auth_usecase_impl.go (hunk 3 above)
Comment (luckyAkbar) [reply internal/app/services/core/auth/auth_usecase_impl.go:461]: Okay, I think now I can implement that. I'll make sure to notify you again after it has been fully tested again

Comment (luckyAkbar) [reply internal/app/services/core/auth/auth_usecase_impl.go:461]: The very first opening the app on a fresh browser (i cleared all the cookies before closing the browser)

image
image


refreshing on home page will trigger anonymous session creation

image


Filling a questionnaire -> go back to homepage and refresh (without claiming the QR result) -> fill a questionnaire again and claim the result will only claim the last QR since the anonymous token was refreshed 

ctx:
current cookie's guest-id: 409fabb5-63ca-4f16-8f86-ad457ea44406

first submitted QR resp (id: DHAYO4EBJED7VDMJ):
image

go back to homepage and refresh there
image


fill another questionnaire (QR id: DHAYPCKRU4G5NNTX):
image

claim the last QR:

image

image


the onlly claimed resources is the last QR after refreshing the homepage

File: internal/app/delivery/http/controllers/auth_controller.go (hunk 2 above)
Comment (luckyAkbar) [reply internal/app/delivery/http/controllers/auth_controller.go:366]: For now I choose not to provide this kind of endpoint mas @lamurian . The reason is, related [to this discussion](https://github.com/konsulin-care/konsulin-api/pull/300#discussion_r2746323044), upon manual refresh in homepage, the app perform hard refresh to existing anonymous tokens, which effectively delete the old one and create a new token. 

If this reason still aligns with your original intent here, I think we can safely resolve this conversation, mas

Comment (lamurian) [reply internal/app/delivery/http/controllers/auth_controller.go:366]: Suppose that we want to add other token-deletion triggers in the future, I assume we can use similar approach (hard refresh existing tokens) as an action to that trigger? If so, then it resolves the concerns above, and we don't need an extra endpoint.

Comment (luckyAkbar) [reply internal/app/delivery/http/controllers/auth_controller.go:366]: yes, that can be achieved by the current anonymous session endpoint, mas. I also predict that we can also easily extend the endpoint if in the future we need to embed more information to the token that might be useful using the same approach (extending the capabilities using query parameters / body request)

**Review by lamurian — APPROVED** [review]
